azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -331,78 +338,211 @@ def __init__(self, xml_text):
-                  "{1} but found versions: ({2})".format(version, name, ', '.join(
-                set([getattrib(x, "version") for x in ext_handler_plugin_settings]))) 
+            msg = "ExtHandler PluginSettings Version Mismatch! Expected PluginSettings version: {0} for Handler: {1} but found versions: ({2})".format(
+                version, handler_name, ', '.join(set([getattrib(x, "version") for x in ext_handler_plugin_settings])))
             add_event(AGENT_NAME, op=WALAEventOperation.PluginSettingsVersionMismatch, message=msg, log_event=False,
                       is_success=False)
-            if not settings:
-                # If there is no corresponding settings for the specific extension handler, we will not process it at all,
-                # this is an unexpected error as we always expect both versions to be in sync.
-                logger.error(msg)
-                return
-            logger.warn(msg)
-
-        runtime_settings = None
-        runtime_settings_node = find(settings[0], "RuntimeSettings")
-        seqNo = getattrib(runtime_settings_node, "seqNo")  # pylint: disable=C0103
-        runtime_settings_str = gettext(runtime_settings_node)
-        try:
-            runtime_settings = json.loads(runtime_settings_str)
-        except ValueError as e:  # pylint: disable=W0612,C0103
-            logger.error("Invalid extension settings")
-            return
+            raise ExtensionConfigError(msg)
+
+        if len(settings) > 1:
+            msg = "Multiple plugin settings found for the same handler: {0} and version: {1} (Expected: 1; Available: {2})".format(
+                handler_name, version, len(settings))
+            raise ExtensionConfigError(msg)
+
+        plugin_settings_node = settings[0]
+        runtime_settings_nodes = findall(plugin_settings_node, "RuntimeSettings")
+        extension_runtime_settings_nodes = findall(plugin_settings_node, "ExtensionRuntimeSettings")
+
+        if runtime_settings_nodes and extension_runtime_settings_nodes:
+            # There can only be a single RuntimeSettings node or multiple ExtensionRuntimeSettings nodes per Plugin
+            msg = "Both RuntimeSettings and ExtensionRuntimeSettings found for the same handler: {0} and version: {1}".format(
+                handler_name, version)
+            raise ExtensionConfigError(msg)
+
+        if runtime_settings_nodes:
+            if len(runtime_settings_nodes) > 1:
+                msg = "Multiple RuntimeSettings found for the same handler: {0} and version: {1} (Expected: 1; Available: {2})".format(
+                    handler_name, version, len(runtime_settings_nodes))
+                raise ExtensionConfigError(msg)
+            # Only Runtime settings available, parse that
+            ExtensionsConfig.__parse_runtime_settings(plugin_settings_node, runtime_settings_nodes[0], handler_name,
+                                                      ext_handler)
+        else:
```

Comment:
May need to handle this case: both runtime_settings_nodes and extension_runtime_settings_nodes are null.

Reply:
Good point. That is still a valid case right? Like no settings available but its still a new goalstate?

Reply:
yes, for winGA this case is possible.

Reply:
BGInfo for example

Reply:
Ok makes sense, I'll add this case too. Thanks for the catch!

Reply:
I missed it earlier, but we already handle this case in Line 370 :)
Though I still added the condition here too, just to make the code more explicit of what its doing rather than having some implicit dependencies